gtsam/discrete/TableFactor.cpp

```diff
@@ -0,0 +1,340 @@
+  for (Eigen::SparseVector<double>::InnerIterator it(f.sparse_table_); it;
+       ++it) {
+    DiscreteValues assignment_f;
+    for (Key key : f.keys_) {
+      assignment_f[key] = f.lazy_cp(key, it.index());
+    }
+    std::vector<DiscreteValues> projections = project(assignment_f);
+    for (DiscreteValues projection : projections) {
+      size_t m_index = multiplied_f.findIndex(projection);
+      multiplied_f.sparse_table_.insert(m_index) =
+          it.value() * operator()(projection);
+    }
+  }
+  return multiplied_f;
+}
+
+/* ************************************************************************ */
+DecisionTreeFactor TableFactor::toDecisionTreeFactor() const {
+  DiscreteKeys dkeys = discreteKeys();
+  std::vector<double> table;
+  for (size_t i = 0; i < sparse_table_.size(); i++) {
+    table.push_back(sparse_table_.coeff(i));
+  }
+  DecisionTreeFactor f(dkeys, table);
+  return f;
+}
+
+/* ************************************************************************ */
+TableFactor::shared_ptr TableFactor::sum(size_t nrFrontals) const {
+  // Create map to keep track of summed values
+  std::map<size_t, double> cs;
+  size_t cardinality = cardinalities_.at(nrFrontals);
+  for (size_t i = 0; i < cardinality; i++) cs[i] = 0;
+  DiscreteKey target_key = discreteKey(nrFrontals);
+  for (Eigen::SparseVector<double>::InnerIterator it(sparse_table_); it; ++it) {
+    cs[lazy_cp(target_key.first, it.index())] += it.value();
+  }
+  std::vector<double> result;
+  for (std::map<size_t, double>::iterator it = cs.begin(); it != cs.end();
+       ++it) {
+    result.push_back(it->second);
+  }
+  return boost::make_shared<TableFactor>(target_key, result);
+}
+
+/* ************************************************************************ */
+TableFactor TableFactor::max(size_t nrFrontals) const {
+  if (nrFrontals > size())
+    throw invalid_argument(
+        (boost::format("TableFactor::max: invalid number of frontal "
+                       "keys %d, nr.keys=%d") %
+         nrFrontals % size())
+            .str());
+  TableFactor result(*this);
+  for (size_t i = 0; i < nrFrontals; i++) {
+    Key j = keys()[i];
+    result = result.eliminate(j);
+  }
+  return result;
+}
+
+/* ************************************************************************ */
+TableFactor TableFactor::max(const Ordering& frontalKeys) const {
+  // Create map to keep track of summed values
+  if (frontalKeys.size() > size())
+    throw invalid_argument(
+        (boost::format("TableFactor::max: invalid number of frontal "
+                       "keys %d, nr.keys=%d") %
+         frontalKeys.size() % size())
+            .str());
+  TableFactor result(*this);
+  for (size_t i = 0; i < frontalKeys.size(); i++) {
+    Key j = frontalKeys[i];
+    result = result.eliminate(j);
+  }
+  return result;
+}
+
+/* ************************************************************************ */
+TableFactor TableFactor::eliminate(const Key key) const {
```

Comment:
We prefer calling "up" i.e. in this case eliminate should be defined (and declared in .h) *before* max.